File: TestCsharp/tres.cs
/*
 * EJERCICIO:
 * - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
 * - Utiliza operaciones de inserción, borrado, actualización y ordenación.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea una agenda de contactos por terminal.
 * - Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
 * - Cada contacto debe tener un nombre y un número de teléfono.
 * - El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación
 *   los datos necesarios para llevarla a cabo.
 * - El programa no puede dejar introducir números de teléfono no númericos y con más de 11 dígitos.
 *   (o el número de dígitos que quieras)
 * - También se debe proponer una operación de finalización del programa.
 */

using System;

namespace Program
{
    class Estructuras  
    {
        // Funciones de la agenda
        static void insertar(Dictionary<string, int> diccionario)
        {
            Console.WriteLine("Ingrese el nombre del contacto: ");
            string key = Console.ReadLine();
            try
            {
                Console.WriteLine("Ingrese el numero telefonico del contacto: ");
                string value = Console.ReadLine();
                if (value.Length < 11)
                {
                    diccionario.Add(key, Convert.ToInt32(value));
                } else
                {
                    Console.WriteLine("El numero telefonico no puede tener mas de 11 digitos");
                }
            } catch (FormatException e)
            {
                Console.WriteLine("El numero telefonico debe ser un numero entero");
            } catch (ArgumentException e)
            {
                Console.WriteLine("El contacto ya existe");
            }
        }

        static void borrar(Dictionary<string, int> diccionario)
        {
            Console.WriteLine("Ingrese el nombre del contacto a eliminar: ");
            string key = Console.ReadLine();
            try
            {
                
                diccionario.Remove(key);
                Console.WriteLine("Contacto eliminado");
            } catch (Exception e)
            {
                Console.WriteLine("El contacto no existe");
            }
        }

        static void actualizar(Dictionary<string, int> diccionario)
        {
            Console.WriteLine("Ingrese el nombre del contacto a actualizar: ");
            string key = Console.ReadLine();
            Console.WriteLine("Ingrese el nuevo numero telefonico del contacto: ");
            string value = Console.ReadLine();
            try
            {
                if (value.Length > 11)
                {
                    Console.WriteLine("El numero telefonico no puede tener mas de 11 digitos");
                    return;
                } else
                {
                   diccionario[key] = Convert.ToInt32(value);
                   Console.WriteLine("Contacto actualizado"); 
                }
            } catch (FormatException e)
            {
                Console.WriteLine("El numero telefonico debe ser un numero entero");
            } catch (KeyNotFoundException e)
            {
                Console.WriteLine("El contacto no existe");
            }
        }

        static int buscar(Dictionary<string, int> diccionario)
        {
            Console.WriteLine("Ingrese el nombre del contacto a buscar: ");
            string key = Console.ReadLine();
            foreach (KeyValuePair<string, int> kvp in diccionario)
            {
                if (kvp.Key == key)
                {
                    Console.WriteLine("Nombre= {0}, Numero Telefonico = {1}", kvp.Key, kvp.Value);
                    return 0;
                }
            }
            Console.WriteLine("El contacto no existe");
            return 0;
        }

        static void mostrar(Dictionary<string, int> diccionario)
        {
            Console.WriteLine("Contactos:");
            foreach (KeyValuePair<string, int> kvp in diccionario)
            {
                Console.WriteLine("Nombre= {0}, Numero Telefonico = {1}", kvp.Key, kvp.Value);
            }
        }

        // Funcion para el ejemplo de lista
        static void print(List<int> lista)
        {
            foreach (int i in lista)
            {
                Console.WriteLine(i);
            }
        }

        static void Main(string[] args)
        {
            // Las principales estructuras en la programacion son:
            // - Arreglos
            // - Listas
            // - Diccionarios
            // - Colas
            // - Pilas
            // - Conjuntos
            // - Matrices

            // Arreglos 
            int[] arreglo = new int[5];
            arreglo[0] = 1;
            arreglo[1] = 2;
            arreglo[2] = 3;
            arreglo[3] = 4;
            arreglo[4] = 5;

            Console.WriteLine("Arreglo: "); // Leer
            for (int i = 0; i < arreglo.Length; i++)
            {
                Console.WriteLine(arreglo[i]);
            }
            // Listas
            List<int> lista = new List<int>();

            // Insercion, borrado, actualizacion y ordenacion

            lista.Add(1); // Insertar
            lista.Add(2);
            lista.Add(3);
            lista.Add(4);
            Console.WriteLine("Insertar Lista:"); // Leer
            print(lista);

            lista.Remove(2); // Borrar
            Console.WriteLine("Borrar Lista:"); // Leer
            print(lista);

            lista[0] = 5; // Actualizar
            Console.WriteLine("Actualizar Lista:"); // Leer

            print(lista);

            lista.Sort(); // Ordenar
            Console.WriteLine("Ordenar Lista:"); // Leer
            print(lista);

            // Diccionarios

            Dictionary<string, int> diccionario = new Dictionary<string, int>();

            diccionario.Add("uno", 1); // Insertar
            diccionario.Add("dos", 2);

            Console.WriteLine("Insertar Diccionario:"); // Leer
            foreach (KeyValuePair<string, int> kvp in diccionario)
            {
                Console.WriteLine("Key = {0}, Value = {1}", kvp.Key, kvp.Value);
            }

            diccionario.Remove("uno"); // Borrar
            Console.WriteLine("Borrar Diccionario:"); // Leer
            foreach (KeyValuePair<string, int> kvp in diccionario)
            {
                Console.WriteLine("Key = {0}, Value = {1}", kvp.Key, kvp.Value);
            }

            diccionario["dos"] = 5; // Actualizar

            Console.WriteLine("Actualizar Diccionario:"); // Leer   
            foreach (KeyValuePair<string, int> kvp in diccionario)
            {
                Console.WriteLine("Key = {0}, Value = {1}", kvp.Key, kvp.Value);
            }

            // Colas

            Queue<int> cola = new Queue<int>();

            cola.Enqueue(1); // Insertar
            cola.Enqueue(2);

            Console.WriteLine("Insertar Cola:"); // Leer
            foreach (int i in cola)
            {
                Console.WriteLine(i);
            }

            cola.Dequeue(); // Borrar
            Console.WriteLine("Borrar Cola:"); // Leer
            foreach (int i in cola)
            {
                Console.WriteLine(i);
            }

            // Pilas

            Stack<int> pila = new Stack<int>();

            pila.Push(1); // Insertar
            pila.Push(2);

            Console.WriteLine("Insertar Pila:"); // Leer
            foreach (int i in pila)
            {
                Console.WriteLine(i);
            }

            pila.Pop(); // Borrar
            Console.WriteLine("Borrar Pila:"); // Leer
            foreach (int i in pila)
            {
                Console.WriteLine(i);
            }

            // Conjuntos

            HashSet<int> conjunto = new HashSet<int>();

            conjunto.Add(1); // Insertar
            conjunto.Add(2);
            conjunto.Add(3);
            conjunto.Add(2);

            Console.WriteLine("Insertar Conjunto:"); // Leer
            foreach (int i in conjunto)
            {
                Console.WriteLine(i);
            }

            conjunto.Remove(1); // Borrar
            Console.WriteLine("Borrar Conjunto:"); // Leer
            foreach (int i in conjunto)
            {
                Console.WriteLine(i);
            }

            // Matrices

            int[,] matriz = new int[2, 2];

            matriz[0, 0] = 1; // Insertar
            matriz[0, 1] = 2;
            matriz[1, 0] = 3;
            matriz[1, 1] = 4;

            Console.WriteLine("Insertar Matriz:"); // Leer
            for (int i = 0; i < 2; i++)
            {
                for (int j = 0; j < 2; j++)
                {
                    Console.WriteLine(matriz[i, j]);
                }
            }

            // Dificultad Extra

            Dictionary<string, int> agenda = new Dictionary<string, int>();
            int opcion = 0;
            do 
            {
                Console.WriteLine("Agenda de Contactos:");
                Console.WriteLine("1. Insertar Contacto");
                Console.WriteLine("2. Buscar Contacto");
                Console.WriteLine("3. Actualizar Contacto");
                Console.WriteLine("4. Eliminar Contacto");
                Console.WriteLine("5. Mostrar Contactos");
                Console.WriteLine("6. Salir");
                Console.WriteLine("Seleccione una opcion: ");

                try
                {
                    opcion = Convert.ToInt32(Console.ReadLine());
                    if (opcion > 6 || opcion < 1)
                    {
                        Console.WriteLine("Opcion no valida, debe ser un numero entre 1 y 5");
                    } else if (opcion == 1)
                    {
                        insertar(agenda);
                    } else if (opcion == 2)
                    {
                        buscar(agenda);
                    } else if (opcion == 3)
                    {
                        actualizar(agenda);
                    } else if (opcion == 4)
                    {
                        borrar(agenda);
                    } else if (opcion == 5)
                    {
                        mostrar(agenda);
                    } else 
                    {
                        Console.WriteLine("Saliendo");
                    }
                } catch (Exception e)
                {
                    Console.WriteLine("Opcion no valida, debe ser un numero entero");
                }
            } 
            while (opcion != 6);
        }
    }

}

File: TestJava/tres.java
/*
* EJERCICIO:
* - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
* - Utiliza operaciones de inserción, borrado, actualización y ordenación.
*
* DIFICULTAD EXTRA (opcional):
* Crea una agenda de contactos por terminal.
* - Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
* - Cada contacto debe tener un nombre y un número de teléfono.
* - El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación
*   los datos necesarios para llevarla a cabo.
* - El programa no puede dejar introducir números de teléfono no númericos y con más de 11 dígitos.
*   (o el número de dígitos que quieras)
* - También se debe proponer una operación de finalización del programa.
*/
import java.util.Collections;
import java.util.Set;
import java.util.HashSet;
import java.util.Deque;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Dictionary;
import java.util.Enumeration;
import java.util.Hashtable;
import java.util.Queue;
import java.util.Scanner;

class tres {

    static void buscar(Dictionary<String, Integer> agenda, Scanner scanner) {
        System.out.println("Introduzca el nombre del contacto a buscar: ");
        String nombre = scanner.nextLine();

        if (agenda.get(nombre) != null) {
            System.out.println("El contacto " + nombre + " tiene el número " + agenda.get(nombre));
        } else {
            System.out.println("El contacto " + nombre + " no existe");
        }

        System.out.println("-------------");
    }

    static void insertar(Dictionary<String, Integer> agenda,  Scanner scanner) {
        System.out.println("Ingrese el nombre del contacto: ");
        String nombre = scanner.nextLine();

        System.out.println("Ingrese el número de teléfono: ");
        try {
            String number = scanner.nextLine();
            if (number.length() > 11) {
                System.out.println("El número de teléfono no puede tener más de 11 dígitos");
            } else {
                int telefono = Integer.parseInt(number);
                agenda.put(nombre, telefono);
            }
        } catch (NumberFormatException e) {
            System.out.println("El número de teléfono no es valido");
        }
        System.out.println("-------------");
    }

    static void actualizar(Dictionary<String, Integer> agenda, Scanner scanner) {
        System.out.println("Ingrese el contacto a actualizar: ");
        String nombre = scanner.nextLine();

        try {
            if (agenda.get(nombre) != null) {
                System.out.println("Ingrese el nuevo número de teléfono: ");
                String number = scanner.nextLine();
                if (number.length() > 11) {
                    System.out.println("El número de teléfono no puede tener más de 11 dígitos");
                } else {
                    int telefono = Integer.parseInt(number);
                    agenda.put(nombre, telefono);
                }
            } else {
                System.out.println("El contacto " + nombre + " no existe");
            }
        } catch (NumberFormatException e) {
            System.out.println("El numero ingresado no es valido ");            
        } 
        System.out.println("-------------");
    }

    static void eliminar(Dictionary<String, Integer> agenda, Scanner scanner) {
        System.out.println("Ingrese el contacto a eliminar: ");
        String nombre = scanner.nextLine();

        if (agenda.get(nombre) != null) {
            agenda.remove(nombre);
            System.out.println("El contacto " + nombre + " ha sido eliminado");
        } else {
            System.out.println("El contacto " + nombre + " no existe");
        }
        System.out.println("-------------");
    }
    
    static void mostrar(Dictionary<String, Integer> agenda) {
        System.out.println("Mostrando contactos: ");
        for (Enumeration<String> k = agenda.keys(); k.hasMoreElements();) {
            String key = k.nextElement();
            System.out.println("Nombre: " + key + " - Teléfono: " + agenda.get(key));
        }
        System.out.println("-------------");
    }

    static void agenda() {
        // Crear una agenda
        String entrada = "";
        int i = 0;
        Scanner sc = new Scanner(System.in);
        System.out.println("Agenda de contactos: ");
        Dictionary<String, Integer> agenda = new Hashtable<>(); 
        do {
            try {
                System.out.println("1. Buscar contacto");
                System.out.println("2. Insertar contacto");
                System.out.println("3. Actualizar contacto");
                System.out.println("4. Eliminar contacto");
                System.out.println("5. Mostrar contactos");
                System.out.println("6. Salir");

                System.out.println("Introduzca una opción: ");
                entrada = sc.nextLine();
                i = Integer.parseInt(entrada);
                switch(i) {
                    case 1:
                        buscar(agenda, sc);
                        break;
                    case 2:
                        insertar(agenda, sc);
                        break;
                    case 3:
                        actualizar(agenda, sc);
                        break;
                    case 4:
                        eliminar(agenda, sc);
                    break;
                    case 5:
                        mostrar(agenda);
                        break;
                    case 6:
                        System.out.println("Salir");
                        break;
                    default:
                        System.out.println("Error: Opción no válida");
                        break;
                }
            } catch (NumberFormatException e) {
                System.out.println("Error: No es un entero");
            }
            
        } while(i != 6);
        sc.close();
    }
    public static void main(String[] args) {
        // Las principales estructuras en la programacion son:
        // - Arreglos
        // - Listas
        // - Diccionarios
        // - Colas
        // - Pilas
        // - Conjuntos
        // - Matrices
        
        // Arreglos:
        int[] arreglo = {1, 2, 3, 4, 5};
        System.out.println("Arreglo: ");
        for (int i = 0; i < arreglo.length; i++) {
            System.out.println(arreglo[i]);
        }
        System.out.println("-------------");

        // Listas:
        // Listas en Java son implementadas por la clase ArrayList
        // Se pueden agregar, eliminar y actualizar elementos
        // Se pueden ordenar
        // Se pueden recorrer
        // Se pueden buscar elementos
        // Se pueden obtener elementos por índice
        // Se pueden obtener el tamaño de la lista

        // Crear una lista e insertar elemento
        ArrayList<String> lista = new ArrayList<String>();
        lista.add("Elemento 1");
        lista.add("Elemento 3");
        lista.add("Elemento 2");
        lista.add("Elemento 4");
        // Eliminar un elemento
        lista.remove("Elemento 2");

        // Actualizar un elemento
        lista.set(1, "Elemento 5");

        // Recorrer la lista
        System.out.println("Lista: ");
        for (String elemento : lista) {
            System.out.println(elemento);
        }
        System.out.println("-------------");
        
        // Ordenar la lista
        Collections.sort(lista);
        System.out.println("Lista ordenada: ");
        for (String elemento : lista) {
            System.out.println(elemento);
        }

        System.out.println("-------------");

        // Diccionarios:
        // Diccionarios en Java son implementados por la clase Hashtable
        // Se pueden agregar, eliminar y actualizar elementos

        // Crear un diccionario e insertar elementos
        Dictionary<String, Integer> diccionario = new Hashtable<>();
        diccionario.put("Elemento 1", 1);
        diccionario.put("Elemento 2", 2);
        diccionario.put("Elemento 3", 3);
        diccionario.put("Elemento 4", 4);

        // Eliminar un elemento
        diccionario.remove("Elemento 2");

        // Actualizar un elemento
        diccionario.put("Elemento 3", 5);

        // Recorrer el diccionario
        System.out.println("Diccionario: ");
        for (Enumeration<String> k = diccionario.keys(); k.hasMoreElements();) {
            String key = k.nextElement();
            System.out.println("Key: " + key + " - Value: " + diccionario.get(key));
        }


        // Buscar un elemento 
        if (diccionario.get("Elemento 1") != null) {
            System.out.println("El elemento 1 tiene el valor " + diccionario.get("Elemento 1"));
        } else {
            System.out.println("El elemento 1 no existe");
        }   

        System.out.println("-------------");

        // Colas
        // Colas en Java son implementadas por la clase LinkedList
        // Se pueden agregar y eliminar elementos
        // Se pueden recorrer

        // Crear una cola e insertar elementos
        Queue<String> cola = new LinkedList<>();
        cola.add("Elemento 1");
        cola.add("Elemento 2");
        cola.add("Elemento 3");
        cola.add("Elemento 4");

        // Eliminar un elemento
        cola.remove();

        // Recorrer la cola
        System.out.println("Cola: ");
        for (String elemento : cola) {
            System.out.println(elemento);
        }
        System.out.println("-------------");
        // Pila
        // Pilas en Java son implementadas por la clase Stack
        // Se pueden agregar y eliminar elementos
        // Se pueden recorrer

        // Crear una pila e insertar elementos
        Deque<String> pila = new ArrayDeque<>();
        pila.push("Elemento 1");
        pila.push("Elemento 2");
        pila.push("Elemento 3");
        pila.push("Elemento 4");

        // Eliminar un elemento
        pila.pop();

        // Recorrer la pila
        System.out.println("Pila: ");
        for (String elemento : pila) {
            System.out.println(elemento);
        }

        System.out.println("-------------");

        // Conjuntos
        // Conjuntos en Java son implementados por la clase HashSet
        // Se pueden agregar y eliminar elementos

        // Crear un conjunto e insertar elementos
        
        Set<String> conjunto = new HashSet<>();
        System.out.println("Conjunto:");
        conjunto.add("Elemento 10");
        conjunto.add("Elemento 12");
        conjunto.add("Elemento 11");
        conjunto.add("Elemento 12");
        conjunto.add("Elemento 13");

        // Eliminar
        conjunto.remove("Elemento 10");

        for (String elemento : conjunto) {
            System.out.println(elemento);
        }

        // Dificultad extra
        agenda();
    }
}